### ISSUE TITLE
Add a config option to specify that certain URIs are forced to be "anarchist" URIs

### ISSUE THREAD

**Issue by janvarga** [servo/rust-url#488]
We are currently using nsIPrincipal in Gecko to get the origin and base domain information from an URI, but that can only be called on the main thread and we would like to call it directly on a background thread. MozURL which is backed by rust-url can be used on any thread, but it lacks support for some mozilla specific URIs (like moz-extension://foo/bar, resource://foo/bar, about:home, indexeddb://fx-devtools, etc.). Most of those can be handled in MozURL implementation. However there's one thing that has to be checked directly in rust-url parser.

I'm attaching a patch that demonstrates the issue:

[bug_1526891_part_1_add_support_for_simple_uris_to_rust_url_r_nika.txt](https://github.com/servo/rust-url/files/2897729/bug_1526891_part_1_add_support_for_simple_uris_to_rust_url_r_nika.txt)

Basically, those URIs shouldn't have an authority, so some form of configuration to specify that these URIs are forced to be "anarchist" URIs is required.

What would be the best way to support that ?

**Comment by SimonSapin** [comment]
@annevk, do you feel this is a use case that the spec should address or is it better to have an internal flag not exposed to web content that changes the behavior of the parser?

**Comment by annevk** [comment]
Is this style of parsing web-exposed? The idea is that handling of such schemes needs to happen at a higher level than the parser. I.e., some kind of processor that takes the output of the parser and enforces that there's no authority or some such.

Looking at the patch it seems this would be observable through `new URL()`.

**Comment by janvarga** [comment]
I also tried to do that at higher level, but in that case we still need to tell rust-url to treat it as "anarchist URI". Another option would be to have a list of custom schemes that is passed to rust-url, or have a custom function, etc.

**Comment by SimonSapin** [comment]
To be clear, "anarchist" is what a code comment in rust-url calls URLs that can be base URLs (and so have a path with slash-separated segments) but have no authority (no username, password, host, or port component).

@annevk There seems to be an existing use case inside Gecko for URLs like `moz-extension://foo/bar` where `foo` is not a host name but the first segment of a path. Do you think we should try to find a way to tweak the parsing algorithm in the spec (for everyone including web content) to support this use case, or should rust-url instead have an alternate parsing mode (not exposed to web content) for internal use by Gecko?

@janvarga As far as I understand, this issue is only about inputs where the scheme is followed by `://`, so `about:home` is not concerned. For the other cases, would it be possible at all to retcon those URLs so that the first path component becomes an authority (host name), and use rust-url without parser changes? Or, could these URLs be changed to use a single slash after the `:`?

**Comment by janvarga** [comment]
Actually, foo in moz-extension://foo/bar is a host name in Gecko.
foo in indexeddb://foo/bar is the first segment of a path
So given this input: indexeddb://fx-devtools
Origin string in Gecko is: indexeddb://fx-devtools
Origin string in rust-url is: indexeddb://fx-devtools/
The slash is not the only problem, host_str() should return None

All those special schemes (about, app, indexeddb, moz-safe-about) are currently handled by nsSimpleURI in Gecko, which means they don't have a hostname.

It seems that the list can be reduced to just indexeddb (indexeddb://) and app (app://) scheme, rust-url seems to already do the right job for about and moz-safe-about.

We can't change any URLs because we use origin and base domain string for accessing storage on disk. It must stay as it is and MozURL must 100% match what Gecko currently returns if we want to use it for quota manager, etc.

@SimonSapin Does that answer all your questions ?

**Comment by annevk** [comment]
Why can't you abstract the way origin is computed and do it differently for these schemes?

@SimonSapin getting agreement on changing the URL parser is quite hard. Chrome does have a similar broken architecture, but they'll use different schemes, so I'm not sure what good that does us.

**Comment by janvarga** [comment]
Here's how we compute origin in nsIPrincipal for indexeddb://fx-devtools
origin = uri.asciiSpec
which results in indexeddb://fx-devtools

Here's what MozURL can do
origin = url.as_ref() or url.as_str() (both should be equivalents for uri.asciiSpec)
which results in indexeddb://fx-devtools/ (note the trailing slash)

I could add code that removes the slash, but then if there's input like this:
indexeddb://fx-devtools/
nsIPrincipal computes: indexeddb://fx-devtools/
MozURL computes: indexeddb://fx-devtools

I could do even more hackery, but I think it would be cleaner to treat that URI as URI with no authority. The host name would be empty too which gives better correctness (not just right origin string).

**Comment by annevk** [comment]
Thank you, I see what you mean. If you want to distinguish `indexeddb://fx-devtools` and `indexeddb://fx-devtools/` there is indeed no way to do that without changing the URL parser.

**Comment by SimonSapin** [comment]
@janvarga I’m a bit confused as to which URLs you want to have a host name or not, your last comment seems to contradict the patch posted in the first one. And based on [nsIURI.idl](https://searchfox.org/mozilla-central/rev/00f3836a87b844b5e4bc82f698c559b9966e4be2/netwerk/base/nsIURI.idl#212-217) it looks like `asciiSpec` is a serialization of an entire URL, so using that as an origin seems wrong.

Taking a step back a bit: the spec is designed to support any scheme/protocol that one might come up with while requiring as little as knowledge as possible in the parser about those different schemes. When a parsed string starts with a scheme followed by `:`, there are three cases:

* The `:` is followed by `//`: the URL has a host name and a path which contains slash-separated segments. When serialized, the path always starts with `/`.
* The `:` is followed by a single `/`: the URL does not have a host name, that slash marks the start of a path like in the host name case.
* The `:` is not followed by any slash: what follows (until the query string or fragment, if any) is an arbitrary string that may or may not contain slashes, but those slashes are not significant. There is no notion of a relative URL. (The spec calls this `cannot-be-a-base`.) That component is exposed in the JS API as `path` but it is not "really" a path.

What would be cleaner in my opinion is if Firefox-internal URLs could be made to conform to one of these cases, so that no parser change (deviation from the spec) is necessary.

If that is not practical, a parser change should only be introduced in a way that doesn’t affect web content: add a new field and setter method to the `ParseOptions` struct that enables an alternative parsing mode. This mode should only be enabled for internal uses not exposed to the web.
